### PR TITLE
support for load/save of dht storage state

### PR DESCRIPTION
**Pull request by aldenml**
What do you think about this idea?

### CODE REVIEW

File: include/libtorrent/kademlia/dht_storage.hpp [hunk 1]
```diff
@@ -216,6 +216,10 @@ namespace dht
 
 		virtual dht_storage_counters counters() const = 0;
 
+		virtual void load_state(std::vector<char> state) = 0;
```
Comment (aldenml) [inline include/libtorrent/kademlia/dht_storage.hpp]: missing const& here

File: src/kademlia/dht_state.cpp [hunk 2]
```diff
@@ -81,6 +90,13 @@ namespace
 			ret.nodes6 = detail::read_endpoint_list<udp::endpoint>(nodes);
 #endif
 
+		if (bdecode_node const st = e.dict_find_string("storage-state"))
+		{
+			char const* ptr = st.string_ptr();
```
Comment (arvidn) [inline src/kademlia/dht_state.cpp]: this is meant to be a "blob" defined by the storage implementation? would there need to be protection from passing in the blob from another storage class?

**Comment by arvidn** [comment]
What kind of data would you expect a storage implementation to save here? What it stores internally is quite well defined, right? it can only be peer lists and data blobs

File: src/kademlia/dht_state.cpp (hunk 2 above)
Comment (aldenml) [inline src/kademlia/dht_state.cpp]: Yes, it will be up to the implementors protect from faulty "blob"s.

**Comment by aldenml** [comment]
rebased for a complete review

**Comment by aldenml** [comment]
I understand this is a kind of weak API. Do you have an idea of how to improve this?

**Comment by aldenml** [comment]
Closing this for a fresh start

**Comment by arvidn** [comment]
sorry for not responding. I've been kind of busy the last week.

**Comment by aldenml** [comment]
please, no need :)

I do want to have something robust that clients can use by default, but I don't think it's a good idea to have internal types in that "state" like the `bloom_filter`

**Pull request closed without merging**